### PR TITLE
RGW - Add Multipart and Versioned POSIX

### PR DESCRIPTION
**Pull request by dang**
This is a rework of the POSIXDriver.  It refactors out the actual posix parts into a set of classes that provide access to underlying directory/file/symlink, etc.  These primatives are used to build up full support for Bucket, Object, Multipart, and VersionedObjects.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/driver/posix/bucket_cache.h [hunk 1]
```diff
@@ -472,12 +478,12 @@ struct BucketCache : public Notifiable
     using namespace LMDBSafe;
 
     int rc{0};
-    GetBucketResult gbr = get_bucket(bname, BucketCache<D, B>::FLAG_LOCK);
+    GetBucketResult gbr = get_bucket(nullptr, bname, BucketCache<D, B>::FLAG_LOCK);
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/bucket_cache.h:481]: nullptr is the prefix provider?

File: src/rgw/driver/posix/bucket_cache.h [hunk 2]
```diff
@@ -544,6 +550,78 @@ struct BucketCache : public Notifiable
...
+  } /* add_entry */
+
+  int remove_entry(const DoutPrefixProvider* dpp, std::string bname, cls_rgw_obj_key key) {
+    using namespace LMDBSafe;
+
+    GetBucketResult gbr = get_bucket(dpp, bname, BucketCache<D, B>::FLAG_LOCK);
+    auto [b /* BucketCacheEntry */, flags] = gbr;
+    if (b) {
+      unique_lock ulk{b->mtx, std::adopt_lock};
+      ulk.unlock();
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/bucket_cache.h:594]: is it correct unlock before the lmdb txn?

Comment (mattbenjamin) [reply src/rgw/driver/posix/bucket_cache.h:594]: (compare l. 620)

File: src/rgw/driver/posix/rgw_sal_posix.cc [hunk 3]
```diff
@@ -19,45 +19,156 @@
...
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator &bl) {
+    DECODE_START(1, bl);
+    decode(user, bl);
+    decode(display_name, bl);
+    DECODE_FINISH(bl);
+  }
+  friend inline std::ostream &operator<<(std::ostream &out,
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/rgw_sal_posix.cc:63]: if it's inline here why is it a friend?

File: src/rgw/driver/posix/rgw_sal_posix.cc [hunk 4]
```diff
@@ -19,45 +19,156 @@
...
+static inline std::string gen_rand_instance_name()
+{
+  enum { OBJ_INSTANCE_LEN = 32 };
+  char buf[OBJ_INSTANCE_LEN + 1];
+
+#if 0
+  gen_rand_alphanumeric_no_underscore(driver->ctx(), buf, OBJ_INSTANCE_LEN);
+#else
+  static uint64_t last_id = UINT64_MAX;
+  snprintf(buf, OBJ_INSTANCE_LEN, "%lx", last_id);
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/rgw_sal_posix.cc:98]: fmt::format()?

File: src/rgw/driver/posix/rgw_sal_posix.cc [hunk 5]
```diff
@@ -115,121 +226,1629 @@ static int get_x_attrs(optional_yield y, const DoutPrefixProvider* dpp, int fd,
...
+    ret = del->remove(dpp, null_yield, /*delete_children=*/true);
+    if (ret < 0) {
+      ldpp_dout(dpp, 0) << "ERROR: could not remove cur_version " << get_name()
+                        << dendl;
+      return ret;
+    }
+  }
+
+  /* Create new current version symlink */
+  /* XXX dang do in writer::complet? */
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/rgw_sal_posix.cc:1495]: anything left to do here?

File: src/rgw/driver/posix/rgw_sal_posix.cc [hunk 6]
```diff
@@ -422,12 +2034,7 @@ std::unique_ptr<Notification> POSIXDriver::get_notification(
 
 int POSIXDriver::close()
 {
-  if (root_fd < 0) {
-    return 0;
-  }
-
-  ::close(root_fd);
-  root_fd = -1;
+  //root_dir->close();
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/rgw_sal_posix.cc:2037]: what's going on here (commented code)?

File: src/rgw/driver/posix/rgw_sal_posix.cc [hunk 7]
```diff
@@ -982,7 +2547,7 @@ int POSIXBucket::put_info(const DoutPrefixProvider* dpp, bool exclusive, ceph::r
   struct timespec ts[2];
   ts[0].tv_nsec = UTIME_OMIT;
   ts[1] = ceph::real_clock::to_timespec(mtime);
-  int ret = utimensat(parent_fd, get_fname().c_str(), ts, AT_SYMLINK_NOFOLLOW);
+  int ret = utimensat(dir->get_parent()->get_fd(), get_fname().c_str(), ts, AT_SYMLINK_NOFOLLOW);
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/rgw_sal_posix.cc:2550]: cool

**Review by mattbenjamin — APPROVED** [review]
tests could better describe themselves, maybe;  looks very good

File: src/rgw/driver/posix/bucket_cache.h (hunk 1 above)
Comment (dang) [reply src/rgw/driver/posix/bucket_cache.h:481]: correct

File: src/rgw/driver/posix/bucket_cache.h (hunk 2 above)
Comment (dang) [reply src/rgw/driver/posix/bucket_cache.h:594]: This is how notify() works, and I copied that.  It gets the lock, unlocks, then creates and commits the transaction.  Maybe the lock isn't necessary for this bit at all?

File: src/rgw/driver/posix/rgw_sal_posix.cc (hunk 4 above)
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:98]: Is fmt::format() better than snprintf()?  There's *lots* of C code in here that will have to be replaced when we switch to aio.

File: src/rgw/driver/posix/rgw_sal_posix.cc (hunk 5 above)
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:1495]: Nope.  Comment removed.

File: src/rgw/driver/posix/rgw_sal_posix.cc (hunk 6 above)
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:2037]: Left over from when the FDs were in the main SAL classes.  Unused, and removed.

**Comment by dang** [comment]
jenkins test api

**Comment by dang** [comment]
jenkins test api

**Comment by dang** [comment]
jenkins test api

**Comment by dang** [comment]
jenkins test make check
jenkins test api

**Comment by dang** [comment]
jenkins test make check

**Comment by dang** [comment]
jenkins test api

**Comment by dang** [comment]
jenkins test make check

**Comment by dang** [comment]
jenkins test make check

**Comment by dang** [comment]
jenkins test make check

**Pull request merged**